### ISSUE TITLE
Redefine the project

### ISSUE THREAD

**Issue by matthiasbeyer** [matthiasbeyer/imag#56]
This issue was re-used for other purposes

Here goes the discussion on redefining the project.

---

**The old post:**

```
The store should be git version controlled.

The following things should be considered:

- [ ] opt-out - git should be on by default, causing imag to crash if git is not available on the system
- [ ] Commandline
  - [ ] option to specify the commit message on a `git-commit`
  - [ ] option to open the `$EDITOR` on `git-commit`
    - [ ] editor should be either configured in the config file (`commit-editor`) or the default editor for `imag` (which automatically falls back on`vim` if there is none)
- [ ] use `git-rm` for file removal
- [ ] Verify that there are no unstaged files, refuse store action if there are any
- [ ] Commit messages should be configured in the configuration file, optionally with format-strings, for
  - [ ] Adding a new file to the store
  - [ ] Editing a file in the store
  - [ ] Removing a file from the store
  - [ ] The format string should be able to embed:
    - [ ] The module of the file beeing altered
    - [ ] The id (hashtype, hash) of the file beeing altered
    - [ ] An arbitrary string, dependend of the module of the file beeing edited. For example the URL for the bookmark module, the name of a note for the note module
  - [ ] Commandline-Call should be included in the commit message body
  - [ ] Default commit messages should be available, hard-coded but also in the configuration file example which should exist at some point
- [ ] ... more?

---

Another killer-feature (opt-in) would be: If several files beeing edited, the store backend should open a new branch for this, commit each file change individually and merge this branch back into `master` afterwards.

This would require some tense refactoring, as we do not have an option to `persist()` several files in one step in the storage backend by now, but it should be achievable.
```

**Comment by neithernut** [comment]
In a discussion @matthiasbeyer and I had on IRC, some issues were discussed regarding git as a storage facility for imag. I'd like to revise these issues as well as options. As soon as we made a decision, it should be incorporated into the design document.

### Tracking of content

It was discussed whether the content of an item (e.g. a file referenced by an item) should be tracked by git. While this seems most natural for some items (e.g. notes), which are expected to be bundled with the metadata in some way anyway, for other items it's not as ideal (e.g. media-files). This may be either due to the fact that the file is expected to live outside the directory in which the repository resides or the type of the file itself (you don't want binary files in your git repository).

I'd like to discuss the following options:
 * For content which is not entangled with the metadata, an URI (or multiple URIs) reference the file. There are no additional references.
 * The file is referenced by a tree object (just a file checked into a repository). The item references the file/content by it's sha1-sum. This option would probably require some quirks to reliably address files outside the actual repository, but bring other benefits.
 * The file is "imported": a copy is added to the repository. An URI stored within the metadata references the original. This is/may be essentially the same mechanism used for content which is bundled with the metadata by default.

There are pros and cons for each option dependent on the type of data and the use case. I'd like to propose an cascading option which let the user choose one behaviour globally/for each module/for each item. The user may set the option globally. Module-specific options override the global option. Lastly, a command-line-option overrides both the global and the module-specific option.

The behaviour is set for each item created. For existing items, it should be trivial to detect the behaviour selected from the circumstances (e.g. the presence of an URI and the nature or absence of the in-repository reference). In this case, the behaviour should be preserved for an item unless the user specifies otherwise via the command line.

Most naturally, we should carefully define sane default behaviours for each module.


### Metadata/Item storage format

@matthiasbeyer pointed out that, at least when using current libraries, we may want to stick to standard git object types. When doing so, metadata could be stored as a blob, formatted in a standardized way (the current choices proposed by @matthiasbeyer are json and yaml). Personally, I don't think it's that much of a problem to use blobs to store metadata. How to organize them is, however, another issue.

Ideally, metadata blobs would just float in the emptiness of space, but they have to be attached to something to (a) prevent deletion on runs of the garbage collector, (b) make them actually findable or at least list-able and (c) ultimately link them to a commit. This is normally done by attaching them to a tree, which, in turn, is attached to a commit.

Organizing the metadata in a git-tree has a number of consequences:
 * As with directory listings, this typically yields in long lists of entries for a directory, which, on top, is potentially altered with every commit.
 * If the metadata is stored as a blob, it is addressed by a hash. If we are to put that blob into a tree, we may have to come up with a name for the "file" (I didn't yet investigate whether anonymous entries are supported). We could re-use the hash, but this would look kinda dumb.
 * When using a tree, we can divide the metadata-blobs by modules and even come up with some (user-defined) grouping of items (that's an idea from the back of my head which I have yet to formulate in a different issue).

For the contents, using blobs stored in trees is the only sensible option IMHO. Now if we organize the metadata in trees, too, we have a number of options:
 * _If_ anonymous entries are allowed, we could put the (now invisible) metadata next to the contents. However, this would disable any grouping feature which is not bound to the (perceived) location of the file making up the contents.
 * We could make the (two?) trees subtrees of another tree (e.g. have a sub-folder/sub-tree ".metadata"). This would be a viable option, but it may still not be ideal.
 * We could attach both sub-trees to a commit directly. I did not yet investigate whether this is even possible.
 * We could create a "shadow" commit for either metadata or contents for each "front" commit and make the "shadow"-commit a parent of the "front" commit. The shadow-commit may then be used to hold the other tree. This may look funny in the log and in diffs and it may break if people use git-commands on the repository directly, but it might be worth it.
 * We could manage two completely separate refs, e.g. `ref/heads/data` and `ref/heads/metadata`. Needless to say, this would potentially be a primary source of confusion and foot-shooting, especially if people create branches on their own.

Comments?

----

On a side note, I'm not sure whether providing management functionality which essentially mirrors the git front end makes that much sense, especially when considering experiences users which may want to do just use the repository directly. Maybe it would be better to let users run the regular git-commands for syncing/versioning stuff and just work on the index. Optionally, some of the git-commands may be forwarded by imag.
For some special management-functionality, a higher level interface absolutely makes sense. This can, however, be postponed for now, I think.

**Comment by matthiasbeyer** [comment]
All of what you said sounds nice and I agree mostly.

I overslept the whole issue and came to a simple but IMHO complete solution. Let me elaborate:

1. I think there is no format we can define which works for all kind of input. At least not with lots of effort which would kill the whole project.
1. Using the git infrastructure (the git internals) would be nice-ish, but would result in some complexity for both the implementation and the end-user - I wouldn't go for this.
1. I think it is better to _just do it_ instead of talking around. Because of this, I'd like to propose the following changes to the current status:
  * Split the commands as planned into several libraries and frontends for them - ACK. Lets do this! Not that important, though. First: Get shit done.
  *  Putting content in the git object store - Nope, to complex.
  *  Putting everything human-readable into a "store" - ACK
  *  Putting the things as structured data into the "store" - ACK.
  * Keeping the current idea of `file := header + content` - ACK, but transforming the header type to `YAML`-only, as it is user-editable this way
  *  Having a `store/<hash>.ext` path format for content in the store: Nope. Proposal:
    * Each Module has a subdirectory in the store.
    * Under this subdirectory, the user is free to add files, as he would in a normal git repository. Example: `/store/notes/on_cooking/pasta.md`. Doing it this way has several simple advantages:
      * Links are literally paths
      * Links to versions are still possible (e.g.: `my [pasta note](/store/notes/on_cooking/pasta.md @ <sha1 of the version>`)
      * The user is able to use the editor to edit content
      * The user is able to use git to do things without any drawbacks
    * As `1.6.2.4` says: Having the `store` git-version controlled - Yes, like a normal source-code project
  * There still can be core-commands for editing header data of files in a module-independent way
  * The issue with the external content:
    * content which is text could be either 
      * copied to the store (we're talking of KB, maybe MB, not GB) -> version control enabled automatically because store is in git
      * Referenced by path
    * Paths to media-files can be put as simple files: ```yaml
- type: path
- pointer: /some/path/to/a/file.mp3
- sha1sum: <...>
- file-metadata: <...>
```
This way we can re-find the content if it gets "lost".

The overall point is:

* "Start small" (quote by Linus Torvalds, actually)
* Get shit done
* Don't talk too much, just do it.

@TheNeikos , @neithernut Comments welcome

---

@neithernut I'm not rejecting your comment with this post. I'm proposing an alternative which (IMHO) seems way more simple and flexible (thinking of implementation as well as usability).

**Comment by neithernut** [comment]
The difference between out proposals actually isn't that great. In a sense, yours can be seen as a subset of mine (with concrete choices where I posted options), with the exception that you want to stick to the "bundle header and content into one file" idea, which _is_ reasonable. I totally agree with your proposal as long as there is reasonable room for extension. When putting stuff into a `store` subdirectory, do have the freedom we need for extension.

One annotation I have is that it would be a good idea to let the user choose whether content should be referenced by path or copied to the store (as in my proposal).

When choosing this model, there is also no need for actually requiring git. IMHO we should let the user choose whether she wants to make the thing a repository or not.

----

This is off-topic: I agree that starting small is almost always a good idea (if you leave some room for extension) and that you can talk things dead. However, when "just doing it" you may get caught up in a dead end. Weeks of work may be for naught. So thinking about this things first does pay out, especially when we're talking about the very core of the system.

Nonetheless, we may just implement your proposal and just check whether it actually pays out (which I think it does) or not.

**Comment by matthiasbeyer** [comment]
> One annotation I have is that it would be a good idea to let the user choose whether content should be referenced by path or copied to the store (as in my proposal).

ACK. Defaults would be: Text content will be copied by default, binary referenced.

> When choosing this model, there is also no need for actually requiring git. IMHO we should let the user choose whether she wants to make the thing a repository or not.

ACK. I'd enable it by default if git is available on the system.

> Nonetheless, we may just implement your proposal and just check whether it actually pays out (which I think it does) or not.

Nice to hear!

---

So, I'd like to know @TheNeikos opinion on the ideas and then we should go ahead and define _simple_ and _small_ issues (which should result in _simple_ and _small_ PRs) for transforming the current code-base into what we are aiming for.

**Comment by TheNeikos** [comment]
My opinion to this:

- I like the idea of using git for versioning (nothing else about it strikes me as useful for this though) of metadata.
- Use TOML instead of YAML, it is imo functionally easier to use, has great Rust integration, does not need excessive whitespace.

Otherwise I do not see anything that strikes me completely odd,

**Comment by matthiasbeyer** [comment]
> TOML

I completely agree. We should remove the JSON/YAML support completely for this and integrate TOML tightly in my opinion.

**Comment by neithernut** [comment]
Just read up TOML
So it's essentially INI with some additional feature. Like it.

**Comment by neithernut** [comment]
Btw: I'd like to see an option to enable/disable use of git (e.g. when creating the store), but this should be an easy-to-make addition once the system is proven to work for us.

**Comment by matthiasbeyer** [comment]
So, lets define issues 

> we should go ahead and define simple and small issues (which should result in simple and small PRs) for transforming the current code-base into what we are aiming for.

@neithernut , @TheNeikos Do you guys have ideas where to start?

**Comment by matthiasbeyer** [comment]
Some ideas:

- [ ] Remove JSON Parser
- [ ] Remove YAML Parser
- [ ] Redefine Store-Interface
  - [ ] Library
    - [ ] Create
    - [ ] Read
    - [ ] Update
    - [ ] Delete
  - [ ] CLI
  - [ ] TOML parser
  - [ ] File Format
    - [ ] Header
    - [ ] Content
      - [ ] Markdown-Parser
      - [ ] replaceable with other text formats
- [ ] Define Runtime library (`librt`), providing 
  - [ ] CLI-Flags
    - [ ] Verbosity
    - [ ] Debugging
  - [ ] Configuration (config-file)
  - [ ] Store Interface
  - [ ] ... The `Runtime` type in the current codebase is basically what we want here
- [ ] Define Module-generic library (`libmodule`), providing traits for
  - [ ] Get CLI-spec from module
  - [ ] Get name from module
  - [ ] Call/Execute module with CLI values

Basically we have the following dependency structure:

```
libcore <- libstore <- librt <- libmodule
```

`librt` and `libmodule` also depend on `libcore`, not only implicitely.
`libcore` holds utilities.
`librt` contains configuration-file and (basic) CLI parsing
`libmodule` defines a high-level module interface.

`libstore` (and maybe `libmodule`) will get a CLI frontend later.

**Comment by neithernut** [comment]
You listed CLI in both the store library _and_ the runtime library. This doesn't make much sense. I would remove the CLI from the store interface.
For clarification: everything listed under "Redefine Store-Interface" ultimately goes into `libstore`? Or should the basic file format handling go into the utility library?
Also, I assume that you listed the points in preferred order of implementation?

**Comment by TheNeikos** [comment]
If you mean that the store module self should not have a CLI then I agree, it should go into one of the others

**Comment by neithernut** [comment]
Yep, exactly.

**Comment by neithernut** [comment]
Btw: maybe @matthiasbeyer should rewrite the front post (or at least make those not checkboxes), so we don't get the non-progressing progression bar in the issue overview.

**Comment by matthiasbeyer** [comment]
Ah, of course we should use the document now as _documentation and definition paper_ of the project and get shit done implementation-wise.

**Comment by neithernut** [comment]
How's the status on this issue now? Should it be closed? If not, what would be the criteria for closing it?

**Comment by matthiasbeyer** [comment]
I'm not sure. I consider it partly resolved, though I'd like to wait until we have some working base. I think I will close it as soon as the store has a working commandline interface, what do you think?

**Comment by neithernut** [comment]
Sounds reasonable.

**Comment by matthiasbeyer** [comment]
I guess I will close this now, as the store binary is implemented and works. Some features of the store are not implemented yet, though the base is working.

@neithernut @TheNeikos let me thank you for your input by now, I think this project is heading into a nice direction.
